subsys/bluetooth/mesh/heartbeat.c

```diff
@@ -409,6 +409,10 @@ void bt_mesh_hb_resume(void)
 static int hb_pub_set(const char *name, size_t len_rd,
 		      settings_read_cb read_cb, void *cb_arg)
 {
+	if (!bt_mesh_is_initialized()) {
```

Comment:
I think this could be hidden under `BT_MESH_SETTINGS_DEFINE`, e.g.:
```
#define BT_MESH_SETTINGS_DEFINE(_hname, _subtree, _set)			     \
	static int pre_ ## _set(const char *name, size_t len_rd, \
		      settings_read_cb read_cb, void *cb_arg) \
	{ \
		if (!atomic_test_bit(bt_mesh.flags, BT_MESH_INIT)) { \
			return 0; \
		} \
		return _set(name, len_rd, read_cb, cb_arg); \
	} \
	SETTINGS_STATIC_HANDLER_DEFINE(bt_mesh_##_hname, "bt/mesh/" _subtree, \
				       NULL, pre_##_set, NULL, NULL)
```

Reply:
However, I would really like to be sure that there is no more neat way of solving this. Have you checked if there is another way to make settings not load subtree? Or maybe we could let settings subsystem recall mesh settings but once we get to `mesh_commit`, check that stack was initialized. Though in this case we have to be sure that nothing dangerous happens in `_set` handlers.

Reply:
A more neat way to solve this might be to pass all settings for bt/bt_mesh trough the bt settings handler, the generic bt handler can then disable all bt_mesh settings handlers. Possibly it could even avoid all bt_mesh settings handlers to be registered.

Reply:
@Laczen could you elaborate on your suggestion? Are talking about calling all mesh set handlers from bt settings set handler instead of registering them through `SETTIGNS_STATIC_HANDLER_DEFINE`?

Reply:
I tried your suggestion. I think because I have to parse subtree inside mesh, it adds little bit more footprint in comparison with flag check in every set handler (tested on `mesh_shell` sample), but this probably can be improved:

Default:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      369292 B         1 MB     35.22%
             RAM:       50508 B       256 KB     19.27%
        IDT_LIST:          0 GB         2 KB      0.00%
```

With pre_##_set:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      369532 B         1 MB     35.24%
             RAM:       50508 B       256 KB     19.27%
        IDT_LIST:          0 GB         2 KB      0.00%
```

With own iterable section (https://github.com/PavelVPV/zephyr/commit/860214998860d839846221f659121499cfbdf24b):
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      369744 B         1 MB     35.26%
             RAM:       53644 B       256 KB     20.46%
        IDT_LIST:          0 GB         2 KB      0.00%
```

Reply:
It will be so much better if the settings_load_subtree() API has another parameter that accepts array of subtree string pointers (acting as filters) that can be included/excluded from loading. Something like:
`int settings_load_subtree(const char *subtree, const char *subtree_filter_list, size_t filter_entries, enum filter_type_t filter_type);`
Where filter_type_t, can specify the inclusion or exclusion filter.

Reply:
Cool, there seems to be quite some higher ram usage. I am sure this can be improved, but at least you have shown how it could be done without large code modification.

Reply:
@omkar3141, that wouldn't be a solution as it is a call to `settings_load()` that is generating this problem. If all systems would limit themselves to a `settings_load_subtree()` of their own subtree the problem would also disappear.

Reply:
One could remove the call to settings_load() from the application and instead directly use settings_load_subtree() as a top-level loading function. However, there can be application-specific concerns (such as this issue) that may require loading only parts of the tree, for example, when separate subsystems are present and only some of them needed to be active. Pushing these calls in the sub-system init sequence is also not an option because that makes the user application inflexible. @PavelVPV's proposal seems like a good compromise for now.

Reply:
@PavelVPV, I think your `pre ## set` solution is the nicest. It could even be extended to taking a test condition as part of the macro (as a future settings improvement).

Maybe handling for a NULL _set routine should be added at the macro level.